src/cargo/sources/git/utils.rs

```diff
@@ -282,6 +297,12 @@ impl<'a> GitCheckout<'a> {
                 .with_checkout(checkout)
                 .fetch_options(fopts)
                 .clone(url.as_str(), into)?;
+            if database.repo.is_shallow() {
```

Comment:
Here we 'upgrade' the local clone capability of `git2` to handle shallow source repos correctly. This might be something to add to `libgit2` or `git2` one day.